utils/read-closest-ts-config-by-path.ts

```diff
@@ -0,0 +1,86 @@
+import type ts from 'typescript'
+
+import * as path from 'node:path'
+import * as fs from 'node:fs'
+
+import { getTypescriptImport } from './get-typescript-import'
+
+// Heavily inspired from https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/src/parseSettings/getProjectConfigFiles.ts
+
+interface InputProps {
+  tsconfigRootDir: string
+  filePath: string
+}
+
+export const directoryCacheByPath = new Map<string, string>()
+export const contentCacheByPath = new Map<string, ts.CompilerOptions>()
+
+export const readClosestTsConfigByPath = (
+  input: InputProps,
+): ts.CompilerOptions | null => {
+  let typescriptImport = getTypescriptImport()
+  if (!typescriptImport) {
+    return null
+  }
+
+  let directory = path.dirname(input.filePath)
+  let checkedDirectories = [directory]
+
+  do {
+    let tsconfigPath = path.join(directory, 'tsconfig.json')
+    let cachedDirectory = directoryCacheByPath.get(directory)
+    if (!cachedDirectory && fs.existsSync(tsconfigPath)) {
+      cachedDirectory = tsconfigPath
+    }
+
+    if (cachedDirectory) {
+      for (let dir of checkedDirectories) {
+        directoryCacheByPath.set(dir, cachedDirectory)
+      }
+      return getCompilerOptions(typescriptImport, cachedDirectory)
+    }
+
+    directory = path.dirname(directory)
+    checkedDirectories.push(directory)
+  } while (
+    directory.length > 1 &&
+    directory.length >= input.tsconfigRootDir.length
+  )
+
+  throw new Error(
+    `Couldn't find any tsconfig.json relative to '${input.filePath}' within '${input.tsconfigRootDir}'.`,
+  )
+}
+
+const getCompilerOptions = (typescriptImport: typeof ts, filePath: string) => {
+  if (contentCacheByPath.has(filePath)) {
+    return contentCacheByPath.get(filePath)!
+  }
+  let configFileRead = typescriptImport.readConfigFile(
+    filePath,
+    typescriptImport.sys.readFile,
+  )
+  if (configFileRead.error) {
+    throw new Error(
+      'Error reading tsconfig file: ' + JSON.stringify(configFileRead.error),
+    )
+  }
+  let parsedContent = typescriptImport.parseJsonConfigFileContent(
+    configFileRead,
+    typescriptImport.sys,
+    './',
```

Comment:
Not fully sure if this is what should be entered here.

![image](https://github.com/user-attachments/assets/947c2b4e-8b53-498f-aab6-64ea16067bb0)

Reply:
It's rather something like `path.dirname(filePath)`

Reply:
@OlivierZal That makes sense, updated: path.dirname(filePath)